### PR TITLE
[Bignum] Bignum cleanup ecp curves

### PR DESCRIPTION
**Pull request by xkqian**
## Description

The recent refactoring of ECP bignum is only needed in the new copy file ecp_curves_new.c. 
The ecp_curves.c will be reverted to the commmit 449bd83 which is before the refactoring work.
Re-apply the changes other than the refactoring work to ecp_curves.c.
Use macro guard to tell which copy is used.

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** ~~provided, or~~ not required(Revert work and no changelog needed)
- [x] **backport** ~~done, or~~ not required
- [x] **tests** ~~provided, or~~ not required (Reuse the current test cases)

### CODE REVIEW

File: tests/suites/test_suite_ecp.function
Comment (minosgalanakis) [inline tests/suites/test_suite_ecp.function:1523]: We may need to update that dependency on the ecp.py [file](https://github.com/Mbed-TLS/mbedtls/blob/development/scripts/mbedtls_dev/ecp.py#L37). ( Multiple occurences )

File: library/ecp_curves.c
Comment (minosgalanakis) [inline library/ecp_curves.c:23]: Any specific reason that this has a double dependency? The code block needs both MBEDTLS_ECP_WITH_MPI_STRUCT and MBEDTLS_ECP_LIGHT to be asserted.

Are we testing possible configuration combination we are generating with the

MBEDTLS_ECP_LIGHT, MBEDTLS_ECP_WITH_MPI_STRUCT, MBEDTLS_ECP_WITH_MPI_UINT, MBEDTLS_ECP_C ?

Comment (xkqian) [reply library/ecp_curves.c:23]: Actually, this PR is based on the PR #7703 , and has not rebased to the latest code.  After updated, there will not be the combination of MBEDTLS_ECP_LIGHT and MBEDTLS_ECP_C .

File: tests/suites/test_suite_ecp.function
Comment (xkqian) [reply tests/suites/test_suite_ecp.function:1531]: Does it mean we should move the test cases into the auto-generated file?

File: tests/suites/test_suite_ecp.function
Comment (paul-elliott-arm) [inline tests/suites/test_suite_ecp.function:1393]: I'm not sure why this has been resurrected - this is now handled by `ecp_mod_p_generic_raw`

**Review by paul-elliott-arm — COMMENTED** [review]
Tests are failing due to ASAN issues - do you need any assistance with this?

**Comment by xkqian** [comment]
> Tests are failing due to ASAN issues - do you need any assistance with this?

Thanks for the reminder, I just want to rebase just after #7758 is merged. Anyway, seems you said #7758 can't be merged cause some preceding dependency, I will rebase it now.

**Comment by paul-elliott-arm** [comment]
We have definite issues here - the rollback of ecp_curves is not possible without corresponding rollbacks of ecp.c (at the very least, I have not tried this). Both ecp_curves.c and ecp.c have newer work that we don't want to remove by the looks of things as well.

**Comment by xkqian** [comment]
> We have definite issues here - the rollback of ecp_curves is not possible without corresponding rollbacks of ecp.c (at the very least, I have not tried this). Both ecp_curves.c and ecp.c have newer work that we don't want to remove by the looks of things as well.

Agree that it's impossible that only rollback ecp_curves.c without rollback of ecp.c. Fortunately we didn't have too much changes for ecp.c since https://github.com/Mbed-TLS/mbedtls/commit/449bd8303eed8164b83682d2ce028dca0e49b1fa , and I find this commit d61dbd4df7deace80bd3e315f6547ba3f925c5c4 is very important about the memory free.  I will revert this commit and hope we can solve the memory leak issues.

**Review by paul-elliott-arm — CHANGES_REQUESTED** [review]
The approach definitely looks good, but my above comments about ecp_mod_p448 in the test suite has not been dealt with, and I still think we should be gating the generated tests on `MBEDTLS_ECP_WITH_MPI_UINT` as Minos pointed out.

**Comment by yanesca** [comment]
The preceding PR has been merged, @xkqian could you please rebase?

**Comment by xkqian** [comment]
> The preceding PR has been merged, @xkqian could you please rebase?

Sure， Rebase done.

**Review by yanesca — DISMISSED** [review]
Did an incremental review on top of @paul-elliott-arm and @minosgalanakis 's review.

LGTM

File: library/ecp_curves.c
Comment (tom-cosgrove-arm) [inline library/ecp_curves.c:5414]: Hmm, I make set bit in nibbles

```
 * with R = 2^32 + 2^12 + 2^8 + 2^7 + 2^6 + 2^3 + 1 = 0x0100001119
              1     1      1        C         9
```

which matches the original - what am I missing here?

File: library/ecp_curves.c
Comment (yanesca) [reply library/ecp_curves.c:4683]: `curve448_part_of_n` is 13818066809895115352007386748515426880336692474882178609894547503885 in hex, whereas `curve448_n` is 2^446 - 13818066809895115352007386748515426880336692474882178609894547503885 in hex.

Comment (tom-cosgrove-arm) [reply library/ecp_curves.c:4683]: And 13818066809895115352007386748515426880336692474882178609894547503885 (decimal) is 8335DC163BB124B65129C96FDE933D8D723A70AADC873D6D54A7BB0D (hex) - thanks for clarifying

File: library/ecp_curves.c
Comment (yanesca) [reply library/ecp_curves.c:5414]: There was a typo in the original, which we fixed during our related work. This development work remains in `ecp_curves_new.c`. This PR is about restoring `ecp_curves.c` to a point prior this work. Keeping improvements that are useful on their own (like this one) is out of scope.

Comment (tom-cosgrove-arm) [reply library/ecp_curves.c:5414]: Okay, so to confirm: you want this to be wrong?

Comment (gilles-peskine-arm) [reply library/ecp_curves.c:5414]: I'm opposed to reverting to known bugs. If we fixed a bug, let's keep the bug fix, in a new commit (but in this pull request) if that's the simplest way.

Comment (yanesca) [reply library/ecp_curves.c:5414]: I am not opposed to fixing this, I am opposed to start looking for similar changes and striving to keep everything useful from the recent bignum work.

Comment (tom-cosgrove-arm) [reply library/ecp_curves.c:5414]: > I am opposed to start looking for similar changes and striving to keep everything useful from the recent bignum work

I don't want you to do that.

I was just surprised to see a change making something actually wrong (I didn't know this was reverting a fix, but I agree with Gilles here)

Comment (xkqian) [reply library/ecp_curves.c:5414]: Actually we only want to revert the code in ecp_curves.c  to the commit https://github.com/Mbed-TLS/mbedtls/commit/449bd8303eed8164b83682d2ce028dca0e49b1fa, and keep the latest refactoring of ecp into the ecp_curves_new.c. So we only focus on the logic error and ignore the comments.    Anyway, this is one comment error obviously, I will update and change it.

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM - thanks for fixing that comment

**Comment by xkqian** [comment]
This change is not related with the DTLS anti-replay, not sure why it failed here: "DTLS proxy: duplicate every packet, server anti-replay off ............. FAIL
  ! pattern 'resend' MUST be present in the Client output
  ! outputs saved to o-XXX-688.log".

**Review by yanesca — APPROVED** [review]
LGTM.

(I have done the rebase and arrived at the same result.)

**Comment by yanesca** [comment]
Just to be pedantic, @tom-cosgrove-arm could you please re-approve after the rebase?

**Pull request merged**
